### PR TITLE
Include charm extra bindings in the application bindings

### PR DESCRIPTION
**Pull request by babbageclunk**
Otherwise after upgrading charms that use them will get errors complaining about missing bindings from `network-get`. They're slightly fiddly to get - since the 1.25 DB doesn't know about them we need to extract the charm zip from storage and parse the metadata.yaml.

### CODE REVIEW

File: juju1/state/migration_export.go [hunk 1]
```diff
@@ -931,6 +944,76 @@ func (e *exporter) addApplication(ctx addApplicationContext) error {
 	return nil
 }
 
+func extraBindings(application *Service) ([]string, error) {
```
Comment (wallyworld) [inline juju1/state/migration_export.go:947]: oh yeah, they used to be services!

File: juju1/state/migration_export.go [hunk 2]
```diff
@@ -931,6 +944,76 @@ func (e *exporter) addApplication(ctx addApplicationContext) error {
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer reader.Close()
+
+	localFile, err := ioutil.TempFile("", "charm-"+ch.URL().Name)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer os.Remove(localFile.Name())
```
Comment (wallyworld) [inline juju1/state/migration_export.go]: just a single defer func with both cleanups in the one func?

File: juju1/state/migration_export.go [hunk 3]
```diff
@@ -931,6 +944,76 @@ func (e *exporter) addApplication(ctx addApplicationContext) error {
...
+
+	// Find metadata.yaml.
+	var metadataFile *zip.File
+	for _, f := range zipFile.File {
+		if f.Name == "metadata.yaml" {
+			metadataFile = f
+		}
+	}
+	if metadataFile == nil {
+		return nil, errors.Errorf("no metadata file in the %q charm!", ch.Meta().Name)
```
Comment (wallyworld) [inline juju1/state/migration_export.go]: Should this be fatal or just a warning? do we really care to halt the whole operation if there's a single dodgy charm? there'd be do extra bindings in that case anyway. Same comment to some of the other errors below, eg the yaml one

**Review by wallyworld — APPROVED** [review]

File: juju1/state/migration_export.go (hunk 2 above)
Comment (babbageclunk) [reply juju1/state/migration_export.go]: Yeah, will do - especially given that I had to check to remember which order the cleanups happen in.

File: juju1/state/migration_export.go (hunk 3 above)
Comment (babbageclunk) [reply juju1/state/migration_export.go]: Yeah, good call - I'll change that.

**Pull request merged**